fs: fix write methods param validation and docs

LiviaMedeiros opened the pull request:
Refs: #41666 

~~Potentially breaking changes.~~

~~An attempt to implement "named parameters" for [filehandle.write](https://nodejs.org/api/fs.html#filehandlewritebuffer-offset-length-position) (Promises API) and [fs.writeSync](https://nodejs.org/api/fs.html#fswritesyncfd-buffer-offset-length-position) (Synchronous API).~~

1. Segregated `ArrayBufferView` and `string` forms in documentation.
Current documentation, also including [fs.write](https://nodejs.org/api/fs.html#filehandlewritebuffer-offset-length-position) (Callback API), described behaviour for string|Object argument in [buffer, offset, length, position] version.
2. Removed Objects with own .toString functions from accepted parameters.
This functionality doesn't help with most cases, when an object inherits .toString from its prototype
This functionality doesn't work with Symbol.toPrimitive
This functionality is easily replaceable with `obj.toString()`, `String(obj)` or `""+obj`
This functionality in Promises and Synchronous APIs seems to be broken for a long time
3. ~~Function signatures are chosen to match their .read counterparts.
Promises API: `buffer, offset, length, position` => `options`
Synchronous API: `buffer, offset, length, position` => `buffer, options`~~

At this point, these changes should not break any existing working code.
For code trying to use documented feature of accepting objects with explicit .toString, outcome changes from abortion to exception.

Any feedback is appreciated.

~~Edit [2022-02-26]:
Soft-blocked by https://github.com/nodejs/node/pull/42128
Default value of `length` is being changed for reading methods: it should be aligned for writing methods, and it would require adjustments in tests here.~~

~~Edit [2022-02-27]:
WIP, adjusting default `length` and tests.~~ done

Edit [2022-04-04]:
This PR doesn't add new feature anymore (https://github.com/nodejs/node/pull/41677/commits/847749fa7c07a0f48a6576a25d601860fde9048e)
Optional params will be added in separate PR.

benjamingr commented:
@nodejs/fs

LiviaMedeiros commented:
Added tests, covered somewhat related edge cases in existing tests, rebased on master branch.

aduh95 commented on test/parallel/test-fs-write-sync-optional-params.js:
```diff
@@ -0,0 +1,112 @@
+const buffer = Buffer.from('zyx');
+
+{
+  let fd = fs.openSync(dest, 'w+');
```
It would be useful to wrap all the testing in `try`/`finally` block so we don't leave open fd if the tests fails. Creating a `test` function could help. Same thing for the async tests.

```suggestion
  let fd
  try {
    fd = fs.openSync(dest, 'w+')
    test(fd);
  } finally {
    if (fd != null) fs.closeSync(fd)
  }
```

LiviaMedeiros replied:
Thanks for suggestion!
I've wrapped the whole "open, test, close" sequence in functions to ensure fresh fd/fh for each test. This will prevent any uncertainty when `typeof position !== 'number'`, so current position won't be affected by previous iterations.

aduh95 commented on test/parallel/test-fs-write-sync-optional-params.js:
```diff
@@ -0,0 +1,112 @@
+    fs.writeSync(fd, {}); // Missing buffer, {} is second argument
+  }, {
+    code: 'ERR_INVALID_ARG_TYPE',
+    name: 'TypeError',
+    message: 'The "buffer" argument must be of type string or an instance ' +
+      'of Buffer, TypedArray, or DataView. Received an instance of Object'
```
We only test for error code, so the test doesn't fail if the error message is changed in the future. Please only keep the `code` property in the other tests as well.

```suggestion
```

LiviaMedeiros replied:
Affirmative, removed redundant properties in ac0e126f482f6cf46aa18d695473d88828ebdd2f

aduh95 reviewed (commented):
We should add entries to the `changes` list in the YAML comment to say that objects with an own `toString` method are no longer supported with this change.

LiviaMedeiros commented:
Sure, added to `changes` via 6f1344b061073b009039ae711cd32d651dbf04d6

This PR makes separation between `.write(string, ...)` and `.write(buffer, ...)` in the docs stronger, so I've added it only in string version; while previous history stays reflected in both.
Does it look correct, or should any other entries be added or removed?

aduh95 commented on doc/api/fs.md:
```diff
@@ -616,11 +613,33 @@ On Linux, positional writes do not work when the file is opened in append mode.
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/41677
+    description: The `string` parameter must be a primitive string.
```
```suggestion
    description: The `string` parameter no longer accept objects with an
                 own `toString` property.
```

aduh95 commented on doc/api/fs.md:
```diff
@@ -616,11 +613,34 @@ On Linux, positional writes do not work when the file is opened in append mode.
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/41677
+    description: The `string` parameter no longer accept objects with an
```
uhr my bad
```suggestion
    description: The `string` parameter no longer accepts objects with an
```

LiviaMedeiros replied:
Fixed for both methods, rebased on master

mmarchini commented:
cc @nodejs/tsc since it's a semver-major

mcollina reviewed (approved):
lgtm

LiviaMedeiros commented:
I would like to hear any thoughts about `fs.write(fd, string[, position[, encoding]], callback)` from Callback API with `typeof string === 'object'`, and its future.

Aside from issues mentioned earlier, the current implicit object stringifying has two more.
- `Symbol.toPrimitive` (even if inherited) has "higher priority" over `toString` (even if own).
Therefore, this method might do a weird thing: asking for explicit own `.toString`, but using inherited `@@toPrimitive` instead.
Worst case scenario: toPrimitive cast added to ancestor object, with different logic than target's toString.
- Third param of the callback function called `string` and defined as string by documentation is actually not a string but the original object, i.e. `string` param passed in fs.write.

Some ugly code showcasing the problematic behaviour:
```js
import * as fs from 'fs';
const OUTFILE = '/tmp/fswriteobj';

function TstCtor(data) {
  this._data = data;
  this._json = JSON.stringify(data);
  this[Symbol.toPrimitive] = () => this._data?.length;
  this.toString = () => this._json;
}

class TstClass {
  [Symbol.toPrimitive]() { return 'P' };
  toString() { return 'S' };
}

const tst1 = {toString: () => tst1.undefinedProperty};
const tst2 = {toString: () => 'S', [Symbol.toPrimitive]: () => 'P'};

const tst3 = new String('A');
const tst4 = new String('A');
  tst4[Symbol.toPrimitive] = () => 'P';
const tst5 = new String('A');
  tst5.toString = () => 'S';
const tst6 = new String('A');
  tst6[Symbol.toPrimitive] = () => 'P';
  tst6.toString = () => 'S';

const tst7 = new TstCtor(["one","two"]);

const tst8 = new TstClass();
const tst9 = new TstClass();
  tst9.toString = () => 'S';

// "undefined", "P", TypeError, TypeError, "S", "P", "2", TypeError, "P"
fs.open(OUTFILE, 'w', (err, fd) => {
  for (const writeMe of [tst1,tst2,tst3,tst4,tst5,tst6,tst7,tst8,tst9]) {
    try {
      fs.write(fd, writeMe, (err, written, string) => console.log(typeof string));
    } catch(e) {
      console.error(e.name);
    }
  }
});
```
It also shows that toString defined in class doesn't allow typecasting here, since it's not an ownProperty even on direct class instance.

Of course, the second issue can be fixed in docs; and there are several trivial ways to "fix" first one, e.g. by
```patch
- const str = String(buffer);
+ const str = String(buffer.toString());
```
but I don't see any really good solution. It's either breaking changes in API anyway, or documenting on how and why `fs.write(fd, stringableObj, ...)`, `fs.write(fd, String(stringableObj), ...)` and `fs.write(fd, stringableObj.toString(), ...)` might produce different results. Or both.

Am I missing something, or it deserves to become deprecated for now, in favor of `fs.write(fd, options, callback)` form later?

aduh95 commented:
> * `Symbol.toPrimitive` (even if inherited) has "higher priority" over `toString` (even if own).
>   Therefore, this method might do a weird thing: asking for explicit own `.toString`, but using inherited `@@toPrimitive` instead.

Good catch, I think this have been overlooked in the original implementation (own `toString` check was added in https://github.com/nodejs/node/pull/34993). Calling `toString` seems like a good first workaround that can be backported without breaking changes to current release lines.

A deprecation also makes sense to me, I think asking folks to wrap their object in `String()` before passing it to `fs.write` is an OK tradeoff.

LiviaMedeiros commented:
Thanks; I'm leaning towards deprecation for the following reasons:

- I don't see any workaround that will provide reasonable results with strict, obvious and laconic rules.
In any code where result depends on Symbol and meeting the "has own toString" requirement was unintended or obsoleted, things will break. And it won't be fun to debug, since `objectToWrite === stringPassedToCallback && String(objectToWrite) === String(stringPassedToCallback) && String(objectToWrite) === expectedResult`.
Also it will introduce a weird edge case: technically, `.toString` is allowed to return any primitive type. So we'll have to cast it (which means, `objectToWrite.toString()` is not guaranteed to be equivalent to what was written).
Any other rule combination, such as allowing own `@@toPrimitive` to be an alternative, would also pollute the code, have complicated docs and potentially break someone's code.
- `options` argument as in this PR is supported by [Callback/fs.read](https://nodejs.org/api/fs.html#fsreadfd-options-callback), so it would be natural to implement it for `fs.write` later.
- The mentioned PR was originally based on supporting already deprecated uglify-js v2 and IE8.

Thereby, I'd rather not fix it and definitely not backport, unless there's some software that actually relies on this feature, actually suffers from "bug" in it, and for some reason can't be fixed by explicit typecast.
Legacy code usually wants API to be remain unchanged no matter what.

Since this PR is semver-major and code-wise changes will be the same as for promises and sync versions, could it be done there? Or I should open a standalone deprecation PR?

LiviaMedeiros commented:
Adjusted default `length` value according to new behaviour (https://github.com/nodejs/node/commit/246227f4bca1afe06af5a766fc41e251b7d3700a) of reading counterparts, rebased on master branch.

Docfix in `filehandle.write(buffer[, offset[, length[, position]]])` is intentional as it matches the actual code.

aduh95 commented:
@LiviaMedeiros can you please rebase this to fix the git conflict? I'd like to have this landed in time for Node.js 18.0.0.

LiviaMedeiros commented:
@aduh95 Sure, retested&rebased it.

@alsoamit Sorry for ping. If you don't mind, changes from https://github.com/nodejs/node/pull/42518 will be included here as https://github.com/nodejs/node/pull/41677/commits/8d901bef70b9a33b1ed1909a8aa196a4b56c9365
In current master branch, optional chaining is required due to possibility of `buffer == null`. After this PR, `buffer` at that line is guaranteed to be either ArrayBufferView or primitive string, so `?` is indeed redundant.
If you prefer to commit it later by yourself, please let me know and I'll remove it from here.

aduh95 commented on doc/api/fs.md:
```diff
@@ -586,11 +586,11 @@ changes:
 -->
 
-* `buffer` {Buffer|TypedArray|DataView|string|Object}
+* `buffer` {Buffer|TypedArray|DataView}
```
We probably need to add an entry to the changes list for this method as well.

LiviaMedeiros replied:
Yup, rebasing~

aduh95 replied:
We'd still want an entry status for this `filehandle.write` I think https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L589-L597

LiviaMedeiros replied:
Ugh sorry, I've misread the signature.

This one is separated from `fh.write(string, ...)` in docs, so I consider string-related behaviours to be irrelevant here. The mentioned change is just a doc-only correction. I.e. it doesn't matter if first param is `string` or `string|Object`, because these params belong to "another method" anyway.

I think it would make more sense to keep the `v14.12.0 | The string parameter will stringify an object with an explicit toString function.` entry for `fh.write(string, ...)` method, but omit it in `fh.write(buffer, ...)` versions to avoid confusion.
Does it sound better?

aduh95 replied:
Yes I think that would work – as long as the history doesn't imply that it's still supported, we should be good.

aduh95 replied:
We also want to update these ones I think: 
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4442-L4450
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4469
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4526-L4529
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4555
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L5746-L5749
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L5767

LiviaMedeiros replied:
Eventually. Coercion in `fs.write`, `fs.writeFile` and `fs.writeFileSync` (as well as `fs.appendFile`, `fs.appendFileSync`) is doc-only deprecated ([DEP0162](https://github.com/nodejs/node/commit/4b198ffa7e681824dea10ed372a9ed7201c2c7f5)) atm; so these will be updated later.

aduh95 replied:
If we want to make that a runtime deprecation for v18.x, we should hurry, the semver cutoff is next week.

LiviaMedeiros replied:
Affirmative. I can't promise anything right now, but will try to do it in time.

aduh95 reviewed (commented):
Can you also please update the docs for `filehandle.appendFile` please? https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L198-L199

aduh95 commented on doc/api/fs.md:
```diff
@@ -4390,7 +4415,7 @@ changes:
 
 * `fd` {integer}
-* `buffer` {Buffer|TypedArray|DataView|string|Object}
+* `buffer` {Buffer|TypedArray|DataView}
```
We'd need a change entry here too: https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L4354-L4367

LiviaMedeiros replied:
Resolved in the same way as [filehandle.write](https://github.com/nodejs/node/pull/41677#discussion_r841283211): redundant changelog entry is removed, string-related changes are kept in `fs.write(fd, string, ...)` method.

Note: string version of this method *does* coerce Objects even after this PR. This is doc-only segregation.

aduh95 commented on doc/api/fs.md:
```diff
@@ -5791,14 +5815,28 @@ changes:
 
 * `fd` {integer}
-* `buffer` {Buffer|TypedArray|DataView|string|Object}
+* `buffer` {Buffer|TypedArray|DataView}
```
Here as well https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L5806-L5814

LiviaMedeiros replied:
Resolved in the same way as [filehandle.write](https://github.com/nodejs/node/pull/41677#discussion_r841283211): redundant changelog entry is removed, string-related changes were already added for `fs.writeSync(fd, string, ...)` version.

aduh95 commented:
Removing https://github.com/nodejs/node/labels/semver-major label as it seems this PR currently doesn't introduce any breaking change, does it? AFAICT, it only fixes the docs and fixes a few test cases to ensure the process doesn't abort.
Ping @benjamingr as you were the one that added the label.

aduh95 commented on doc/api/fs.md:
```diff
@@ -180,6 +180,10 @@ longer be used.
 <!-- YAML
 added: v10.0.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/41677
+    description: The `data` parameter no longer accepts objects with an
+                 own `toString` function property.
```
In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).

aduh95 commented on doc/api/fs.md:
```diff
@@ -631,11 +627,34 @@ On Linux, positional writes do not work when the file is opened in append mode.
 <!-- YAML
 added: v10.0.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/41677
+    description: The `string` parameter no longer accepts objects with an
+                 own `toString` function property.
   - version: v14.12.0
     pr-url: https://github.com/nodejs/node/pull/34993
     description: The `string` parameter will stringify an object with an
```
In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).
In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).

```suggestion
```

LiviaMedeiros replied:
Yep, all affected methods weren't working with coercion in the first place. Annihilated all related entries.
I initially assumed that "rewriting history" in the docs would be a bad idea; but yes, it is much cleaner this way, and it won't mislead people who use older versions of Node.

> this `fs.md` file is too big

It sure is. 😄 I hope, something like removal of whole Callback API in foreseeable future will reduce it, and both code and docs will become less confusing.

aduh95 commented on doc/api/fs.md:
```diff
@@ -4361,10 +4358,6 @@ changes:
     description: Passing an invalid callback to the `callback` argument
                  now throws `ERR_INVALID_ARG_TYPE` instead of
                  `ERR_INVALID_CALLBACK`.
-  - version: v14.12.0
-    pr-url: https://github.com/nodejs/node/pull/34993
-    description: The `buffer` parameter will stringify an object with an
-                 explicit `toString` function.
```
Hum it looks to me that this does work though, and is still supported 🤔 

```console
$ node -e 'fs.write(1, { toString() { return "string\n" } }, Function.prototype)'
string
```

aduh95 replied:
Arf, I'm still falling for it, I forgot that we document that method twice with different signature, nevermind.

aduh95 commented:
@LiviaMedeiros can you also please remove the documented support in `fsPromises.writeFile` please? https://github.com/nodejs/node/blob/bbf49e7c52b3118e90db550889d0236806cc168a/doc/api/fs.md#L1536-L1539

For reference, here's how I've tested support for those methods:
```console
node -e 'fs.promises.open("/dev/null").then(fh => fh.appendFile({ toString() { return "str\n" } }))' # TypeError
node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "str\n" } }))' # abort
node -e 'fs.promises.open("/dev/null").then(fh => fh.writeFile({ toString() { return "str\n" } }))' # TypeError
node -e 'fs.promises.writeFile("/dev/null", { toString() { return "string" } })' # TypeError
node -e 'fs.writeFile("/dev/null", { toString() { return "string" } }, Function.prototype)' # Works
node -e 'fs.writeFileSync("/dev/null", { toString() { return "string" } })' # Works
node -e 'fs.writeSync(1, { toString() { return "str\n" } })' # TypeError
```

aduh95 commented:
Also, cana we please make the named param another PR? I think we need to have a patch for the `filehandle.write` abort backported up to v14.x ASAP, then we can land the named param as a semver-minor later. Again, sorry for all the back-and-forth on this PR review, if that helps I can offer to work on splitting the PR myself.

aduh95 commented:
Can you also removed remaining documentation for `fsPromise.writeFile` `Object` support at https://github.com/nodejs/node/blob/847749fa7c07a0f48a6576a25d601860fde9048e/doc/api/fs.md#L1524-L1525 https://github.com/nodejs/node/blob/847749fa7c07a0f48a6576a25d601860fde9048e/doc/api/fs.md#L1533-L1535

aduh95 commented:
Landed in 0c9273d1266bbe67fcdc423913fc8c83c259aa83

juanarbol commented:
This should not be landed in v17.x, ref: https://github.com/nodejs/node/pull/42613#discussion_r842465779

aduh95 commented:
> This should not be landed in v17.x, ref: [#42613 (comment)](https://github.com/nodejs/node/pull/42613#discussion_r842465779)

@juanarbol  that’s incorrect, this should land on v17.x.
